# EP Chan Book Implementation

This is a replicate of the strategy shown in the book "Algorithmic Trading Winning Strategies and their Rationale" by Ernest P. Chan. To use this notebook, you need specific virtual environment **qt-env** and setup the ipython kernel as "qt".

Mean reversion is a widely observed phenomenon in finance. Even though individual asset price may not follow mean reversion, but a combination of assets, i.e., a portfolio, may show mean reversion. Therefore, we can constucture a portfolio with mean reversion feature, and then trade on it to gain profit.

To do so, several things need to be done. First of all, we need to test whether a price series is mean reversion. The test is named **ADF test**. The test statistics is $\lambda$/SE($\lambda$), where $\lambda$ is the regression coefficient of equation:

$\Delta y(t) = \lambda y(t-1) +\mu + \beta t+\alpha_1\Delta y(t-1)+\dots+\alpha_k\Delta y(t-k)+\epsilon_t$.

Import necessary modules and print out version information.

In [2]:
import tushare as ts
import pandas as pd
import seaborn as sns

from bokeh.plotting import figure
from bokeh.io import output_notebook, show
from bokeh.models import ColumnDataSource

from datetime import datetime

output_notebook()
pd.show_versions()

Loading BokehJS ...


INSTALLED VERSIONS
------------------
commit: None
python: 2.7.15.final.0
python-bits: 64
OS: Darwin
OS-release: 17.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.23.0
pytest: None
pip: 10.0.1
setuptools: 39.2.0
Cython: 0.28.2
numpy: 1.14.3
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 5.7.0
sphinx: None
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: 4.2.1
bs4: 4.6.0
html5lib: None
sqlalchemy: 1.2.7
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None


The first task is to get a workable dataset. We can do this using *Tushare*.

In [2]:
df = ts.get_hist_data('sh', start='2014-04-30', end='2018-04-30', ktype='M')
df = df.reset_index()
df['date'] = pd.to_datetime(df['date']) # convert str to datetime
print type(df['date'][1])

[Errno 54] Connection reset by peer
<class 'pandas._libs.tslibs.timestamps.Timestamp'>


## Stationary Test

### ADF test

ADF test is included in *statsmodels* module.

In [3]:
from statsmodels.tsa.stattools import adfuller
from bokeh.models import ColumnDataSource

adf_results = adfuller(df['close'])
print adf_results

data = ColumnDataSource(df)
p = figure(x_axis_type='datetime', plot_width=300, plot_height=300)
p.line('date', 'close', line_width=2, source=data)
show(p)

(-3.5952536734996774, 0.005852418336327642, 11, 37, {'5%': -2.9435394610388332, '1%': -3.6209175221605827, '10%': -2.6104002410518627}, 517.7477924492889)


The ADF test statistics is actually $\lambda/\text{SE}(\lambda)$, which is always negative for mean reversion. Here, the ADF test statistics is -3.595, for 95% confidence, we can reject the null hypothesis, that is, the "sh" index is stationary, even though from the plot it does not seem to be stationary.

$\lambda$ indicates the half-life of mean reversion, as half-life$=-log(2)/\lambda$. It can be determined by running a regression fit with $y(t)-y(t-1)$ as the dependent variable and $y(t-1)$ as the independent variable. 

### Hurst Exponent

Hurst exponent $H$ indicates whether the time series is stationary ($H<0.5$), random ($H=0.5$), or trending ($H>0.5$). First, we need to construct a function to [calculate hurst exponent](https://stackoverflow.com/questions/39488806/hurst-exponent-in-python).

In [5]:
from numpy import log10, polyfit, var, subtract

def hurst_ernie_chan(p, lag_range=None):

    p_log = log10(p) # use log price

    variancetau = []
    tau = []
    
    # Create the range of lag values
    if lag_range == None:
        lags = [2]
    else:
        lags = range(2, lag_range) # lag_range < len(ts)

    for lag in lags: 

        #  Write the different lags into a vector to compute a set of tau or lags
        tau.append(lag)

        # call this pp or the price difference
        pp = subtract(p_log[lag:], p_log[:-lag])
        variancetau.append(var(pp))

    # we now have a set of tau or lags and a corresponding set of variances.
    #print tau
    #print variancetau

    # plot the log of those variance against the log of tau and get the slope
    m = polyfit(log10(tau),log10(variancetau),1)

    hurst = m[0] / 2

    return hurst

print "Hurst(sh): %s" %hurst_ernie_chan(df['close'], 5)

Hurst(sh): 0.5088676864098033


### Variance Ratio

Variance ratio can be used to test whether a financial return series is a pure random walk or having some predictability.

In [9]:
from arch.unitroot import VarianceRatio

vr = VarianceRatio(df['close'], 5)
print(vr.summary().as_text())

     Variance-Ratio Test Results     
Test Statistic                  2.055
P-value                         0.040
Lags                                5
-------------------------------------

Computed with overlapping blocks (de-biased)


The possible explanation of above test is as following (from E.P. Chan's book): test statistics equals to 1 means rejection of the random walk and test statistics equals to 0 means it may be a random walk. The p-value gives the smallest probability that you can reject the null hypothesis, which is a random walk. So in this case, with 95% confidence, we can reject the null hypothesis.

### Cointegration

For a pair of assets, we can run a CADF test to determine the cointegration. Let's see how to determine cointegration of two stocks: AAPL and IBM.

Here, we show how to get global market information via *pandas_datareader*.

In [12]:
pd.core.common.is_list_like = pd.api.types.is_list_like # https://stackoverflow.com/questions/50394873/import-pandas-datareader-gives-importerror-cannot-import-name-is-list-like
import pandas_datareader as web
from datetime import datetime

start = datetime(2014, 4, 30)
end = datetime(2018, 4, 30)

apple = web.DataReader('AAPL','iex', start, end) # IEX has stock prices up to 5 years
ibm = web.DataReader('IBM', 'iex', start, end)

5y
5y


Now, let's store the historical price information into database so that we don't need to request from the server every time.

In [15]:
from sqlalchemy import create_engine, MetaData, TEXT, Integer, Float, DateTime, Table, Column

# apple_table.drop(engine)
# ibm_table.drop(engine)

engine = create_engine("mysql+mysqlconnector://root:"+"password"+"@localhost/stocks") # stocks is the name of db
meta = MetaData(bind=engine)

apple_reset = apple.reset_index()
apple_reset.index.name = 'ID'
apple_table = Table('apple', meta,
                 Column('ID', Integer, autoincrement=False),
                 Column('date', DateTime, nullable=False),
                 Column('open', Float, nullable=True),
                 Column('high', Float, nullable=True),
                 Column('close', Float, nullable=True),
                 Column('low', Float, nullable=True),
                 Column('volume', Float, nullable=True),
                 extend_existing=True
                )

ibm_reset = ibm.reset_index()
ibm_reset.index.name = 'ID'
ibm_table = Table('ibm', meta,
                 Column('ID', Integer, autoincrement=False),
                 Column('date', DateTime, nullable=False),
                 Column('open', Float, nullable=True),
                 Column('high', Float, nullable=True),
                 Column('close', Float, nullable=True),
                 Column('low', Float, nullable=True),
                 Column('volume', Float, nullable=True),
                 extend_existing=True
                )

meta.create_all(engine)

apple_reset.to_sql('apple', engine, if_exists='append', index=True)
ibm_reset.to_sql('ibm', engine, if_exists='append', index=True)

Now we can load data directly from database.

In [3]:
from sqlalchemy import create_engine

engine = create_engine("mysql+mysqlconnector://root:"+"password"+"@localhost/stocks") # stocks is the name of db
apple = pd.read_sql_table('apple', engine)
ibm = pd.read_sql_table('ibm', engine)

Once we have the stock prices, we can run a cointegration test (CADF test).

In [4]:
import statsmodels.tsa.stattools as ts

merged = pd.merge(apple, ibm, on = 'date')
x1 = merged['close_x']
y1 = merged['close_y']

cointegration = ts.coint(x1, y1)
print cointegration

(-1.124243958713828, 0.8766385057497718, array([-3.90734883, -3.34220436, -3.0486644 ]))


We found the test statistics is -1.124, greater than the 90% threshold -3.05, therefore, cannot reject the null hypothesis. That is, AAPL and IBM are not cointegration, as expected.

### Bollinger bands

Bollinger bands is a practical mean reversion trading strategy. We enter the trade when price deviates *entryZscore* standard deviations from the mean. And we exit the position when price mean-reverts to *exitZscore* standard deviations from the mean, where *exitZscore* $<$ *entryZscore*.

In this example, we set *entryZscore*=1 and *exitZscore*=0, which means we will exit when the price mean-reverts to the current mean. Mean and standard deviation is calcualted within the lookback period.

Note, rolling OLS is removed from *Pandas*, we can use [another implementation](https://stackoverflow.com/questions/37317727/deprecated-rolling-window-option-in-ols-from-pandas-to-statsmodels) by Brad Solomon. This package works for *Python3*, so we need to make some local modifications in order to work with *Python27*.

**In order to compare results the E.P. Chan's book, from this point onwards, we will use datasets provided by E.P. Chan, and develop the same strategy as described on his book.**

First step is to load .mat data given by E.P. Chan. The .mat data in this case is three dimensional, containing 1500 dates, 67 symbols, and 4 attributes. Therefore, loading it as dataframe requires some extra work.

In [3]:
import scipy.io
import numpy as np
mat = scipy.io.loadmat('../data/inputData_ETF.mat') # load .mat file

""" this part is an attempt to load .mat to dataframe, but failed
however, the operations are useful as a reference
data = pd.DataFrame(data.items()) # convert dict to dataframe
data.columns = ['a', 'b'] # rename columns
data['b'].update(data['b'].apply(lambda l: [item for sublist in l for item in sublist])) # flatten list of lists
data.set_index('a',inplace=True) # use column a as index
data.loc['syms', 'b'] = [item for sublist in data.loc['syms', 'b'] for item in sublist] # flatten row syms and col b
data = data.drop(['__header__', '__globals__', '__version__']) # drop unecessary rows
print data
"""
vol = pd.DataFrame(np.hstack((mat['tday'], mat['vol']))) # use `np.hstack` to make mini dataframe
cl = pd.DataFrame(np.hstack((mat['tday'], mat['cl'])))
lo = pd.DataFrame(np.hstack((mat['tday'], mat['lo'])))
hi = pd.DataFrame(np.hstack((mat['tday'], mat['hi'])))
data = pd.concat([vol, cl, lo, hi], keys=['vol', 'cl', 'lo', 'hi']) # use `pd.concat` to combine mini dataframes
syms = [item for sublist in np.array(mat['syms']) for items in sublist for item in items] # flatten list of lists of lists for symbol names
col_names = ['tday']+syms # prepare the col names for the final dataframe
data.columns = col_names # reset the column names
print data.head()

             tday  BGU  BGZ  BNO  DRN  DRV  EDC  EDZ       EEM  EEV ...   VXZ  \
vol 0  20060426.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  101502.0  NaN ...   NaN   
    1  20060427.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  121329.0  NaN ...   NaN   
    2  20060428.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN   82875.0  NaN ...   NaN   
    3  20060501.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN   72075.0  NaN ...   NaN   
    4  20060502.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN   83598.0  NaN ...   NaN   

           XLB       XLE       XLF      XLI      XLK      XLP      XLV  \
vol 0  22136.0  294271.0   54232.0   8592.0  24503.0   4504.0  21923.0   
    1  83830.0  448795.0  176893.0  14073.0  26474.0   6010.0  44698.0   
    2  65802.0  190539.0  133976.0  28273.0  29100.0  12658.0  24957.0   
    3  21191.0  195958.0  136147.0   7080.0  38514.0   7386.0  10186.0   
    4  33522.0  190415.0   39776.0   3700.0   8762.0  10278.0   7913.0   

           XLY  XME  
vol 0   8179.0  NaN  
    1  21670.0  NaN  
  

In [4]:
# massage the data
temp = data['tday'].apply(lambda x: str(int(x))) # convert float to str
data['tday'].update(temp) # update dataset, now 'tday' is stored as str
temp2 = data['tday'].apply(lambda x: datetime.strptime(x, '%Y%m%d')) # convert str to datetime object
data['tday'] = temp2 # update dataset, now 'tday' is datetime object
print data.head()

            tday  BGU  BGZ  BNO  DRN  DRV  EDC  EDZ       EEM  EEV ...   VXZ  \
vol 0 2006-04-26  NaN  NaN  NaN  NaN  NaN  NaN  NaN  101502.0  NaN ...   NaN   
    1 2006-04-27  NaN  NaN  NaN  NaN  NaN  NaN  NaN  121329.0  NaN ...   NaN   
    2 2006-04-28  NaN  NaN  NaN  NaN  NaN  NaN  NaN   82875.0  NaN ...   NaN   
    3 2006-05-01  NaN  NaN  NaN  NaN  NaN  NaN  NaN   72075.0  NaN ...   NaN   
    4 2006-05-02  NaN  NaN  NaN  NaN  NaN  NaN  NaN   83598.0  NaN ...   NaN   

           XLB       XLE       XLF      XLI      XLK      XLP      XLV  \
vol 0  22136.0  294271.0   54232.0   8592.0  24503.0   4504.0  21923.0   
    1  83830.0  448795.0  176893.0  14073.0  26474.0   6010.0  44698.0   
    2  65802.0  190539.0  133976.0  28273.0  29100.0  12658.0  24957.0   
    3  21191.0  195958.0  136147.0   7080.0  38514.0   7386.0  10186.0   
    4  33522.0  190415.0   39776.0   3700.0   8762.0  10278.0   7913.0   

           XLY  XME  
vol 0   8179.0  NaN  
    1  21670.0  NaN  
    2   

Before doing any strategy, let's plot the close prices of USO and GLD.

In [38]:
data_close_2 = data.xs('cl')[['tday', 'USO', 'GLD']] 
print data_close_2.head()

# create bokeh data object
close_for_plot = ColumnDataSource(data_close_2) # convert pd.DataFrame to bokeh.models.ColumnDataSource

# create a figure object
p = figure(x_axis_type="datetime", title="Closing Prices")
p.grid.grid_line_alpha=0.3
p.xaxis.axis_label = 'Date'
p.yaxis.axis_label = 'Close Price'

# add elements to the figure
p.line('tday', 'USO', color='#A6CEE3', source=close_for_plot, legend='USO price')
p.line('tday', 'GLD', color='#FB9A99', source=close_for_plot, legend='GLD price')
p.legend.location = "top_left"

show(p)

        tday    USO    GLD
0 2006-04-26  69.54  63.65
1 2006-04-27  68.70  62.96
2 2006-04-28  69.62  65.09
3 2006-05-01  71.50  65.16
4 2006-05-02  72.34  66.55


Once we get the dataset straight, we can go ahead develop the strategy. We first need to run a rolling OLS to determine the hedge ratio between the two assets. Then, according to bollinger band definition, we find the trading signal and update the dollar position of each asset class.

In [76]:
from pyfinance.ols import PandasRollingOLS as ols

y_ticker = 'USO'
x_ticker = 'GLD'
y = data[y_ticker].xs('cl')
x = data[x_ticker].xs('cl')

lookback = 20
entryZscore = 1
exitZscore = 0

# use rolling ols to calculate hedge ratios
model = ols(y=y, x=x, window=lookback)
data_lookback = data.xs('cl')[lookback-1:]
betas = model.beta['feature1'] # make sure beta is a pd.Series, same as data_lookback[x_ticker]

# calcluate number of units in the form y-beta*x
yport = pd.DataFrame(data_lookback[y_ticker] - (betas * data_lookback[x_ticker]))
moving_avg = yport.rolling(window=lookback).mean()
moving_std = yport.rolling(window=lookback).std()
Zscore = (yport - moving_avg) / moving_std

# trade signals, boolean
long_entry = Zscore < -entryZscore # buy when price is lower than -entryZscore
long_exit = Zscore >= -exitZscore # sell when price is higher than or equal to -exitZscore
short_entry = Zscore > entryZscore # short buy when price is higher than entryZscore
short_exit = Zscore <= exitZscore # short sell when price is lower than or equal to exitZscore

num_units_long = np.empty((len(yport), 1))
num_units_long = pd.DataFrame(np.where(long_entry, 1, 0))
# num_units_long = pd.DataFrame(np.where(long_exit, 0, np.nan))
# print num_units_long
num_units_short = np.empty((len(yport), 1))
num_units_short = pd.DataFrame(np.where(short_entry, -1, 0))
# num_units_short = pd.DataFrame(np.where(short_exit, 0, np.nan))
num_units = num_units_long + num_units_short
# print num_units

# create dollar position for each asset
temp1 = pd.DataFrame(np.matlib.repmat(num_units, 1, 2))
temp2 = pd.DataFrame(-betas)
temp2['ones'] = np.ones((len(betas)))

data_lookback_2 = pd.concat([data_lookback[y_ticker], data_lookback[x_ticker]], axis=1, sort=False)
position = np.multiply(np.multiply(temp1, temp2), data_lookback_2)
position.columns = [y_ticker, x_ticker]
print position.tail(10)

           USO     GLD
1471 -0.000000    0.00
1472 -0.000000    0.00
1473 -0.000000    0.00
1474 -0.000000    0.00
1475 -3.895685  162.12
1476 -0.000000    0.00
1477 -4.299941  159.89
1478 -6.212288  157.21
1479 -7.124686  158.31
1480 -8.173644  159.37


Next, let's compute the P&L and other performance measures for this strategy.

In [33]:
# daily pnl in dollar
pnl = np.sum(np.multiply(position[:-1], np.divide(np.diff(data_lookback_2, axis=0), data_lookback_2[:-1])), axis=1)

# gross market value
mkt_value = pd.DataFrame.sum(abs(position[:-1]), axis=1)

# return is pnl divided by gross market value
ret = pnl / mkt_value
ret = ret.fillna(method='pad')

# compute Sharpe
sharpe = (np.sqrt(252) * np.mean(ret)) / np.std(ret) # FIX: the answer does not match Chan book
APR = np.prod(1+ret) ** (252.0 / len(ret)) - 1
print('Price spread Sharpe: {:.4}'.format(sharpe))
print('Price spread APR: {:.4%}'.format(APR))

Price spread Sharpe: -1.73
Price spread APR: -32.7996%


Now, let's plot the cumulative returns.

In [75]:
acum_ret = ret.cumsum() # get cumulative return
acum_ret = acum_ret.fillna(method='pad') # fill with pad
acum_ret = acum_ret.fillna(0) # fill the rest NAN with 0

acum_ret_date = pd.concat([data_lookback['tday'][1:].reset_index(drop=True), acum_ret], axis=1) # concat date and acum_ret
acum_ret_date.columns = ['tday', 'acum_ret'] # reset column names

acum_ret_for_plot = ColumnDataSource(acum_ret_date) # convert pd.DataFrame to bokeh.models.ColumnDataSource

p1 = figure(x_axis_type="datetime", title="Cumulative Returns")
p1.grid.grid_line_alpha=0.3
p1.xaxis.axis_label = 'Date'
p1.yaxis.axis_label = 'Cumulative Returns'

p1.line('tday', 'acum_ret', color='#A6CEE3', legend='acum return', source=acum_ret_for_plot)
p1.legend.location = "top_left"

show(p1)

### Kalman filter

Intead of OLS, we can use Kalman filter to find the optimal hedge ratio and moving average. Kalman filter is a linear algorithm to update the expected value of hidden variable based on the latest value of an observable variable. In our case, 

$y(t) = x(t)\beta(t)+\epsilon(t)$,

the asset prices $y$ is the observable variable, whereas $\beta$ is the hidden variable. Note, $\beta$ contains two parts, the slope and intercept, where slope represents the hedge ratio and intercept represents the moving average of the spread.

Kalman filter has been implemented in *pykalman* package.

Read more about Kalman filter in trading: [QuantStart Dynamic Hedge Ratio](https://www.quantstart.com/articles/Dynamic-Hedge-Ratio-Between-ETF-Pairs-Using-the-Kalman-Filter) and [Quantopian Kalman filter lecture](https://www.youtube.com/watch?v=RxIdLu18SsE).

Since data has been loaded and massaged in the last section, we can directly use the dataset.

In [5]:
from pykalman import KalmanFilter

data_kf = data[['tday', 'EWA', 'EWC']]
data_kf_cl = data_kf.xs('cl')
etfs = ['EWA', 'EWC']

def calc_slope_intercept_kalman(etfs, prices):
    """
    Taken directly from QuantStart Dynamic Hedge Ratio website
    Utilise the Kalman Filter from the pyKalman package
    to calculate the slope and intercept of the regressed
    ETF prices.
    """
    delta = 1e-5 # delta is a param to control the transition covariance matrix; delta=0 means linear reg
    trans_cov = delta / (1 - delta) * np.eye(2)
    obs_mat = np.vstack(
        [prices[etfs[0]], np.ones(prices[etfs[0]].shape)]
    ).T[:, np.newaxis]

    kf = KalmanFilter(
        n_dim_obs=1,
        n_dim_state=2,
        initial_state_mean=np.zeros(2),
        initial_state_covariance=np.ones((2, 2)),
        transition_matrices=np.eye(2),
        observation_matrices=obs_mat,
        observation_covariance=1.0,
        transition_covariance=trans_cov
    )

    state_means, state_covs = kf.filter(prices[etfs[1]].values)
    return state_means, state_covs

state_means, state_covs = calc_slope_intercept_kalman(etfs, data_kf_cl)
print state_means

/Users/zhizunzhang/.virtualenvs/qt-env/lib/python2.7/site-packages/scipy/linalg/basic.py:1321: RuntimeWarning: internal gelsd driver lwork query error, required iwork dimension not returned. This is likely the result of LAPACK bug 0038, fixed in LAPACK 3.2.2 (released July 21, 2010). Falling back to 'gelss' driver.
  x, resids, rank, s = lstsq(a, b, cond=cond, check_finite=False)


[[1.3375311  1.3375311 ]
 [1.33953995 1.33953476]
 [1.34020571 1.34019542]
 ...
 [1.15051616 1.59479389]
 [1.14905394 1.59467038]
 [1.14760939 1.59449835]]


Once we have state means ($\beta$), we can go ahead plot the optimal hedge ratio and moving average of the spread.

In [14]:
state_means_temp = pd.DataFrame(np.array(state_means))
state_means_df = pd.concat([data_kf_cl['tday'], state_means_temp], axis=1)
state_means_df.columns = ['tday', 'hedge ratio', 'moving avg']

kf_for_plot = ColumnDataSource(state_means_df) # convert pd.DataFrame to bokeh.models.ColumnDataSource

p1 = figure(x_axis_type="datetime", title="Slope/Hedge Ratio")
p1.grid.grid_line_alpha=0.3
p1.xaxis.axis_label = 'Date'
p1.yaxis.axis_label = 'Hedge Ratio'

p1.line('tday', 'hedge ratio', color='#A6CEE3', legend='slope/hedge ratio', source=kf_for_plot)
p1.line('tday', 'moving avg', color='#FB9A99', legend='intercept/moving avg', source=kf_for_plot)
p1.legend.location = "top_left"

show(p1)